### PR TITLE
blog: August community gems

### PR DESCRIPTION
**Pull request by flippedcoder**
This is for the August Community Gems post. The questions and answers are ready for review.

Still need to add:
- the cover image
- link to the forum
- topic for September office hours
- update the gif at the end
- update tags

### CODE REVIEW

File: content/blog/2021-08-31-aug-community-gems.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  - CML
+  - Git
+---
+
+### [Q: Are TOML files supported for storing model metrics and displaying them via `dvc metrics show`?](https://discord.com/channels/485586884165107732/485596304961962003/865974923079319563)
+
+Thanks for the question @naeljaneLiblikas!
+
+DVC does not support TOML files for metrics. TOML files are only used for
+parameters.
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:28]: :eyes: seems odd. Why the half-support?

File: content/blog/2021-08-31-aug-community-gems.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+You have a few options with DVC for this.
+
+- Set up an external cache on your object storage
+  bucket.[Check out this doc to learn more](https://dvc.org/doc/user-guide/managing-external-data#setting-up-an-external-cache)!
+- Track a large dataset in an external location without downloading it locally.
+  [This doc will tell you more](https://dvc.org/doc/command-reference/add#example-transfer-to-remote-storage).
+
+Great question @Ravi Kumar!
+
+### [Q: Is there a recommended way to specify multiple pipelines in DVC?](https://discord.com/channels/485586884165107732/485596304961962003/864230750325047316)
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:41]: the way the question is written here ("specify multiple pipelines in DVC") is inconsistent with the answer "specify pipelines in multiple DVC yaml files." Seems one needs to be changed.

File: content/blog/2021-08-31-aug-community-gems.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+Thanks @Tups!
+
+### [Q: Is there way to allow different pipelines to have common dependencies and outputs in DVC pipelines?](https://discord.com/channels/485586884165107732/563406153334128681/867747202306146335)
+
+Good question @vgodie!
+
+It is possible to have overlapping dependencies, but not overlapping outputs.
+Having overlapping outputs introduces uncertainty into DVC commands, like
+`dvc checkout`.
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:68]: leaving aside implementation difficulties (with e.g. `dvc checkout`)... what would the logical use-case be of overlapping outputs?

File: content/blog/2021-08-31-aug-community-gems.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+### [Q: How does the CML runner functionality to restart workflows if it's been shut down by AWS (e.g. spot instances) work?](https://discord.com/channels/485586884165107732/728693131557732403/862641924200857660)
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:70]:
```suggestion
### [Q: How does the CML runner restart workflows if it's been shut down by AWS (e.g. spot instances)?](https://discord.com/channels/485586884165107732/728693131557732403/862641924200857660)
```

File: content/blog/2021-08-31-aug-community-gems.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+You shouldn't have to do anything. Spot instances sends a `SIGINT` that we
+handle to restart the workflow. We have been supporting graceful shutdown by
+unregistering runers for a while now.
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:74]:
```suggestion
unregistering runners for a while now.
```

File: content/blog/2021-08-31-aug-community-gems.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+The main difference now is that we restart workflows with unfinished jobs.
+
+We know for sure this isn't working in Azure. Azure is not doing a graceful
+shutdown, but a shutdown now, even when stopping the instance within the Azure
+console.
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:80]:
```suggestion
We know for sure this isn't working in Azure. Azure does not do a graceful
shutdown -- even when stopping the instance within the Azure console.
```

is there an appropriate (preferably upstream) issue we can link to?

File: content/blog/2021-08-31-aug-community-gems.md (hunk 4 above)
Comment (casperdcl) [reply content/blog/2021-08-31-aug-community-gems.md:70]: also again the question ("How does it work?") seems at adds with the answer ("Yes is works").

File: content/blog/2021-08-31-aug-community-gems.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+Thanks for such a good question @andee96!
+
+### [Q: Can I change the endpoint defined here <https://github.com/iterative/cml/blob/master/src/utils.js#L56>? Or does `cml-publish` always save the artifacts on this endpoint?](https://discord.com/channels/485586884165107732/728693131557732403/864444303169421322)
+
+Good question @Nwp8nice!
+
+If you use GitLab you can use the `--native` option to upload to GitLab instead.
+
+It would be nice to be able to offer an alternative link so if you're
+interested, a PR would be awesome! :blush:
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:91]: link to issue https://github.com/iterative/cml/issues/291

File: content/blog/2021-08-31-aug-community-gems.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+### [Q: Is CML used for creating the MLOps workflows, like Apache Airflow?](https://discord.com/channels/485586884165107732/728693131557732403/866624571519664128)
+
+This is a really good question @Ravi Kumar!
+
+CML relies on CI/CD engines like GitHub Actions or GitLab CI/CD, but doesn't
+replace them. It's just a lightweight wrapper and not a complete workflow
+ecosystem like Airflow.
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:99]:
```suggestion
CML is intended to augment existing CI/CD engines like GitHub Actions or
GitLab CI/CD, not replace them. It's a lightweight wrapper and not a complete
replacement workflow ecosystem like Airflow. We don't like reinventing working
wheels.
```

File: content/blog/2021-08-31-aug-community-gems.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+### [Q: Does CML have the ability to run long-run instances, e.g. can I run AWS instance via GitHub Actions that works more than 6 hours?](https://discord.com/channels/485586884165107732/728693131557732403/866730530262351873)
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:101]:
```suggestion
### [Q: Does CML have the ability to cope with long-running instances, e.g. launching an AWS instance via GitHub Actions that lasts more than 72 hours?](https://discord.com/channels/485586884165107732/728693131557732403/866730530262351873)
```

File: content/blog/2021-08-31-aug-community-gems.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+You can definitely run it for 6 hours. The GitHub limit is 72 hours for
+self-hosted runners and GitLab has no limit on self-hosted runners.
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:104]:
```suggestion
Once the GitHub Actions limit of 72 hours is reached for self-hosted
runners, CML will handle restarting the Action and reconnecting to
the runner. Meanwhile, on GitLab there is no time limit to circumvent
for self-hosted runners.
```

CC @DavidGOrtega

File: content/blog/2021-08-31-aug-community-gems.md [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+Thanks @sergechuvakin!
+
+---
+
+https://media.giphy.com/media/l0IycQmt79g9XzOWQ/giphy.gif
+
+At our September Office Hours Meetup we will be do something.
```
Comment (casperdcl) [inline content/blog/2021-08-31-aug-community-gems.md:112]:
```suggestion
At our September Office Hours Meetup we will be doing something.
```

Either reads as v. mysterious or v. obvious :laughing:

**Review by casperdcl — CHANGES_REQUESTED** [review]

File: content/blog/2021-08-31-aug-community-gems.md (hunk 9 above)
Comment (DavidGOrtega) [reply content/blog/2021-08-31-aug-community-gems.md:101]: Users must also specify a higher job timeout via 
```
timeout-minutes: 5000
```
in their workflows to avoid the 6 hours limit

Comment (casperdcl) [reply content/blog/2021-08-31-aug-community-gems.md:101]: with a hyperlink to https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes

File: content/blog/2021-08-31-aug-community-gems.md [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+DVC does not support TOML files for metrics. TOML files are only used for
+parameters.
+
+### [Q: How can I use object storage as a data source for MLOps with DVC?](https://discord.com/channels/485586884165107732/485596304961962003/866393535296176148)
+
+You have a few options with DVC for this.
+
+- Set up an external cache on your object storage
+  bucket.[Check out this doc to learn more](https://dvc.org/doc/user-guide/managing-external-data#setting-up-an-external-cache)!
+- Track a large dataset in an external location without downloading it locally.
```
Comment (dberenbaum) [inline content/blog/2021-08-31-aug-community-gems.md:36]: I would rephrase this to something like: "Track a large dataset in an external location to download later." Eventually, the dataset needs to be downloaded locally to be used, so I think this is tricky to explain without being misleading.

File: content/blog/2021-08-31-aug-community-gems.md (hunk 2 above)
Comment (dberenbaum) [reply content/blog/2021-08-31-aug-community-gems.md:41]: I think the recommendation is to keep each pipeline in a separate `dcv.yaml` file and is kind of explained in the last bullet. Maybe that needs to be moved up to the top?

File: content/blog/2021-08-31-aug-community-gems.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  [This doc will tell you more](https://dvc.org/doc/command-reference/add#example-transfer-to-remote-storage).
+
+Great question @Ravi Kumar!
+
+### [Q: Is there a recommended way to specify multiple pipelines in DVC?](https://discord.com/channels/485586884165107732/485596304961962003/864230750325047316)
+
+There are a few different ways to approach this.
```
Comment (dberenbaum) [inline content/blog/2021-08-31-aug-community-gems.md:43]: Are these actually different ways to approach this? They seem more like explanation of different aspects of the same approach (using multiple `dvc.yaml` files).

File: content/blog/2021-08-31-aug-community-gems.md [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+- DVC will process them just the same as if they were one DVC file i.e.
+  dependencies between stages in different `dvc.yaml` files are still respected.
+- Each `dvc.yaml` file will have its own `dvc.lock` file in the same directory.
+- Splitting a `dvc.yaml` file into multiple files is encouraged where there are
+  clear logical groupings between stages. It avoids confusion, improves
+  readability, and shortens commands by avoiding long paths preceding every
+  filename.
+
+If you want to see the rest of the explanation,
+[check out this PR we have up](https://github.com/iterative/dvc.org/issues/2494).
```
Comment (dberenbaum) [inline content/blog/2021-08-31-aug-community-gems.md:56]: Maybe mention that the PR is for a user guide? Otherwise, it reads to me like the functionality may not exist yet and is still in a PR.

File: content/blog/2021-08-31-aug-community-gems.md (hunk 11 above)
Comment (dberenbaum) [reply content/blog/2021-08-31-aug-community-gems.md:112]: 😆 Guessing you are still trying to figure this one out @flippedcoder and @jendefig?

File: content/blog/2021-08-31-aug-community-gems.md (hunk 1 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:28]: Should I open an issue to add TOML support for metrics?

Comment (dberenbaum) [reply content/blog/2021-08-31-aug-community-gems.md:28]: Sure, if you want to open an issue and reference that discord message, that would be great!

File: content/blog/2021-08-31-aug-community-gems.md (hunk 2 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:41]: I made some updates here, so let me know what you think!

Comment (dberenbaum) [reply content/blog/2021-08-31-aug-community-gems.md:41]: Looks good, although you might want to clarify that multiple `dvc.yaml` files are recommended _but not required_ to specify multiple pipelines.

File: content/blog/2021-08-31-aug-community-gems.md (hunk 11 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md]: Yeah, we were still working on that. 😅 But I've got that updated now!

File: content/blog/2021-08-31-aug-community-gems.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+---
+
+### [Q: Are TOML files supported for storing model metrics and displaying them via `dvc metrics show`?](https://discord.com/channels/485586884165107732/485596304961962003/865974923079319563)
+
+Thanks for the question @naeljaneLiblikas!
+
+DVC does not support TOML files for metrics. TOML files are only used for
+parameters.
+
+We do have an [open issue](https://github.com/iterative/dvc/issues/6402) for
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:30]: ... to upvote or keep up ..

or

Please feel free to chime in so that we can prioritize it.

File: content/blog/2021-08-31-aug-community-gems.md [hunk 16]
```diff
@@ -0,0 +1,130 @@
+---
+title: August '21 Community Gems
+date: 2021-08-31
+description: |
+  A roundup of technical Q&A's from the DVC community.
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:5]: is this description relevant?

File: content/blog/2021-08-31-aug-community-gems.md [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+unregistering runners for a while now.
+
+The main difference now is that we restart workflows with unfinished jobs.
+
+We know for sure this isn't working in Azure. Azure does not do a graceful
+shutdown -- even when stopping the instance within the Azure console.
+
+Thanks for such a good question @andee96!
+
+### [Q: Can I change the endpoint defined here <https://github.com/iterative/cml/blob/master/src/utils.js#L56>? Or does `cml-publish` always save the artifacts on this endpoint?](https://discord.com/channels/485586884165107732/728693131557732403/864444303169421322)
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:93]: make `here` a link (it will look better, no need to put a link explicitly like this)

File: content/blog/2021-08-31-aug-community-gems.md (hunk 15 above)
Comment (shcheklein) [reply content/blog/2021-08-31-aug-community-gems.md:30]: `TOML files are only used` -> `TOML files are can be used` or `TOML files are supported ..`

File: content/blog/2021-08-31-aug-community-gems.md [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+Thanks for the question @naeljaneLiblikas!
+
+DVC does not support TOML files for metrics. TOML files are only used for
+parameters.
+
+We do have an [open issue](https://github.com/iterative/dvc/issues/6402) for
+this so you can keep up with any updates.
+
+### [Q: How can I use object storage as a data source for MLOps with DVC?](https://discord.com/channels/485586884165107732/485596304961962003/866393535296176148)
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:33]: Unfortunately, I don't think this a good Q/A. Question sounds generic enough to provide a very different (simpler) answer - "use DVC remotes". Other answers are pretty advanced and/or optimizations. I would recommend replacing this cc @jorgeorpinel - why did we think this question is about these advanced cases? :)

File: content/blog/2021-08-31-aug-community-gems.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+### [Q: Is there way to allow different pipelines to have common dependencies and outputs in DVC pipelines?](https://discord.com/channels/485586884165107732/563406153334128681/867747202306146335)
+
+Good question @vgodie!
+
+It is possible to have overlapping dependencies, but not overlapping outputs.
+Having overlapping outputs introduces uncertainty into DVC commands, like
+`dvc checkout`.
+
+Sometimes people want to have overlapping directory outputs (different stages
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:74]: Could we provide an example? (for the sake of having some code blocks and/or images in this post - it looks a bit too "empty" at the moment?

File: content/blog/2021-08-31-aug-community-gems.md [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+### [Q: How does the CML runner restart workflows if it's been shut down by AWS (e.g. spot instances)?](https://discord.com/channels/485586884165107732/728693131557732403/862641924200857660)
+
+You shouldn't have to do anything. Spot instances sends a `SIGINT` that we
+handle to restart the workflow. We have been supporting graceful shutdown by
+unregistering runners for a while now.
+
+The main difference now is that we restart workflows with unfinished jobs.
+
+We know for sure this isn't working in Azure. Azure does not do a graceful
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:88]: do we have a workaround? ticket? 

or consider removing this, since question is about AWS

File: content/blog/2021-08-31-aug-community-gems.md (hunk 17 above)
Comment (shcheklein) [reply content/blog/2021-08-31-aug-community-gems.md:93]: I'm not sure we need a link to sources to be honest ... we can just rephrase it ... Could I change an end-point that is being used by the `clm-publish` command to save assets ... or something like this.

File: content/blog/2021-08-31-aug-community-gems.md [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+### [Q: Can I change the endpoint defined here <https://github.com/iterative/cml/blob/master/src/utils.js#L56>? Or does `cml-publish` always save the artifacts on this endpoint?](https://discord.com/channels/485586884165107732/728693131557732403/864444303169421322)
+
+Good question @Nwp8nice!
+
+If you use GitLab you can use the `--native` option to upload to GitLab instead.
+
+It would be nice to be able to offer an alternative link so if you're
+interested, a PR for [this issue](https://github.com/iterative/cml/issues/291)
+would be awesome! :blush:
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:101]: :blush: is not rendered

File: content/blog/2021-08-31-aug-community-gems.md (hunk 20 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:88]: I removed it because the question was about AWS.

File: content/blog/2021-08-31-aug-community-gems.md (hunk 18 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:33]: I replaced this Q/A with one about getting CSV files for the experiments table.

File: content/blog/2021-08-31-aug-community-gems.md (hunk 19 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:101]: I'm not quite sure what to put here as a code block, but the new Q/A has a Python code block. Let me if you think that looks good enough. If not, I'll add some code in here somewhere. 🙂

File: content/blog/2021-08-31-aug-community-gems.md [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+title: August '21 Community Gems
+date: 2021-08-31
+description: |
+  A roundup of technical Q&A's from the DVC community.
+  This month: separate DVC pipelines, working with CML,
+  object storage as a data source, and more.
+descriptionLong: |
+  A roundup of technical Q&A's from the DVC community.
+  This month: separate DVC pipelines, working with CML,
+  object storage as a data source, and more.
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:11]: object storage - is it still part of the Q/A this time?

File: content/blog/2021-08-31-aug-community-gems.md [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+picture: 2021-08-31/gems-cover.png
+author: milecia_mcgregor
+commentsUrl: https://discuss.dvc.org/t/august-21-community-gems/45342t5t4egrf4w
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:14]: we need the right link it looks like

File: content/blog/2021-08-31-aug-community-gems.md [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+date: 2021-08-31
+description: |
+  A roundup of technical Q&A's from the DVC community.
+  This month: separate DVC pipelines, working with CML,
+  object storage as a data source, and more.
+descriptionLong: |
+  A roundup of technical Q&A's from the DVC community.
+  This month: separate DVC pipelines, working with CML,
+  object storage as a data source, and more.
+picture: 2021-08-31/gems-cover.png
```
Comment (shcheklein) [inline content/blog/2021-08-31-aug-community-gems.md:12]: we need to update it before we merge I think

File: content/blog/2021-08-31-aug-community-gems.md (hunk 22 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:11]: Ope! Just cleaned that up.

File: content/blog/2021-08-31-aug-community-gems.md (hunk 23 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:14]: Got that link up.

File: content/blog/2021-08-31-aug-community-gems.md (hunk 24 above)
Comment (flippedcoder) [reply content/blog/2021-08-31-aug-community-gems.md:12]: Yep! We're just waiting for Roman to send the image. :)

Comment (jorgeorpinel) [reply content/blog/2021-08-31-aug-community-gems.md:12]: Hi. So is this image the only thing pending? Should we then approve @shcheklein ? Thanks

**Comment by jorgeorpinel** [comment]
Hi @casperdcl I see your requested changes (https://github.com/iterative/dvc.org/pull/2655#pullrequestreview-714078943) are all resolved. Do you now approve? 🙂

**Review by casperdcl — APPROVED** [review]

**Comment by jorgeorpinel** [comment]
@flippedcoder should we merge this? Also feel free to add to https://github.com/orgs/iterative/projects/167?fullscreen=true

File: content/blog/2021-08-31-aug-community-gems.md (hunk 24 above)
Comment (jendefig) [reply content/blog/2021-08-31-aug-community-gems.md:12]: Compressed Image is up.  So I believe this is ready to go, though date is for tomorrow.  @flippedcoder @jorgeorpinel @shcheklein

**Comment by flippedcoder** [comment]
@jorgeorpinel Can we get this merged today?

**Comment by shcheklein** [comment]
@flippedcoder  looks like image is not rendered https://dvc-org-aug-comm-gems-hheuambn.herokuapp.com/blog ?

**Comment by flippedcoder** [comment]
> @flippedcoder looks like image is not rendered https://dvc-org-aug-comm-gems-hheuambn.herokuapp.com/blog ?

@shcheklein Should be good now! I had to update image name and merge with master.

**Comment by shcheklein** [comment]
It's fixed, I see also that URL is not correct (not the same pattern as we use usually), let's fix it please

**Comment by flippedcoder** [comment]
> It's fixed, I see also that URL is not correct (not the same pattern as we use usually), let's fix it please

I updated that url to `aug-21-community-gems` and left out the `dvc` part. Since we usually have questions that cover dvc and cml I think it makes sense to leave that out.

**Comment by shcheklein** [comment]
Sounds good (removing DVC), pattern is still not the same - I think we use full month name for heartbeats and gems. We also need to update the link in the forum I think?

**Comment by flippedcoder** [comment]
> Sounds good (removing DVC), pattern is still not the same - I think we use full month name for heartbeats and gems. We also need to update the link in the forum I think?

Just made that update to the full month name. The link in the forum should be fine once this is live now that the url name is correct.

**Review by shcheklein — APPROVED** [review]

**Comment by shcheklein** [comment]
Thanks @flippedcoder 🎉
